### ISSUE TITLE
Lein packages .DS_Store files in jar

### ISSUE THREAD

**Issue by hlship** [technomancy/leiningen#59]
lein jar will package up .DS_Store files in the project root (along with project.clj).

**Comment by brunchboy** [comment]
I just looked in a jar file created by Leiningen 2.8.1 running on MacOS X High Sierra (10.13.6), and there are indeed two .DS_Store files corresponding to folders under `resources/` which I had visited in the Finder, so this is definitely still an issue. I don’t see any way to reopen the issue, though… should I submit a new one?

**Comment by brunchboy** [comment]
Thanks, what you’ll want to do is open a folder within the `resources/` folder of a project in the Finder, then do something like rearranging the position of icons within Icon view, to make sure the Finder creates the `.DS_Store` file to capture that change. Then when you build a `jar`, you will find the `.DS_Store` file included in the `jar`, which it should not be since it is just UI state for the developer’s Mac.

If it would be easier, I could just supply the `jar` file I created as an example.

**Comment by winks** [comment]
I think it's https://github.com/technomancy/leiningen/blob/master/src/leiningen/jar.clj#L102 - I think that second '#' is just wrong, but maybe I'm not seeing what the regex is supposed to do.. @technomancy - maybe this is an easy fix - I just don't know if it's intended to ignore all dot files in a jar.

**Comment by winks** [comment]
Yeah, I hadn't checked how `skip-file?` was exactly used in the whole codebase, thus my comment about "what it supposed to do" - not sure if there would be some file format starting with `#` - but I'm probably overthinking it and "all dotfiles" are supposed to be matched :P

**Comment by brunchboy** [comment]
Anything I can help with on this?

**Comment by technomancy** [comment]
The docstring is incorrect; that function doesn't exclude dotfiles on its own. However, the default value for `:jar-exclusions` (set in the `leiningen.core.project` namespace) does include a pattern `#"^\."`; this would prevent `.DS_Store` files from making it into the jar.

It looks like `.DS_Store` files are being excluded by default as far as I can tell. If you replace the `:jar-exclusions` setting with your own value, then you will disable that, but that's the intended behavior.

**Comment by technomancy** [comment]
I'm going to close this on out since I can't reproduce, but if you have a test case we can reopen.

**Comment by brunchboy** [comment]
Hold on, I’m reproducing it right now. I’ll let you know what I find.

**Comment by brunchboy** [comment]
I have no `:jar-exclusions` in my `project.clj`, and I just created an überjar which ended up containing six .DS_Store directories:

```
jar tvf target/beat-link-trigger.jar | grep DS_Store
  6148 Fri Oct 26 13:09:34 CDT 2018 user_guide/assets/.DS_Store
  6148 Tue May 24 18:10:04 CDT 2016 images/.DS_Store
  6148 Sat Oct 06 21:17:44 CDT 2018 fonts/Teko/.DS_Store
  8196 Sat Oct 06 21:17:44 CDT 2018 fonts/.DS_Store
  6148 Wed Jun 14 23:45:06 CDT 2017 fonts/Orbitron/.DS_Store
  6148 Tue Oct 02 22:15:10 CDT 2018 beat_link_trigger/.DS_Store
```

**Comment by brunchboy** [comment]
I’m running Leiningen 2.8.1 on Java 1.8.0_181 Java HotSpot(TM) 64-Bit Server VM

The project I am building can be found at https://github.com/brunchboy/beat-link-trigger

**Comment by brunchboy** [comment]
I should also confirm that those six files also appear in the ordinary project `.jar` file that was built in the process of creating the überjar. I’m really frustrated that this keeps getting closed even though the problem remains. Please let me know if there is any other way I can help get it fixed, short of diverting my energy to studying the internals of Leiningen.

**Comment by technomancy** [comment]
I see; it will correctly block `.DS_Store` in the root of the `resources` directory but won't block ones that are nested more deeply. Happy to take a patch which changes the default `:jar-exclusions` to a regex which will catch both.

**Comment by brunchboy** [comment]
Hmm, it would seem safer to apply the regular expression to the non-directory component of the filename—or at least to have a mechanism that does that even if it uses a different key—rather than trying to build an expression that knows how to recognize such things on all filesystems.

**Comment by brunchboy** [comment]
I guess one could be built using `java.io.File/pathSeparator`, although that would require a dynamic regular expression, rather than one built from a string literal.

**Comment by brunchboy** [comment]
I can take a crack at this…

**Comment by technomancy** [comment]
> I can take a crack at this…

Great; thanks.

> Hmm, it would seem safer to apply the regular expression to the non-directory component of the filename

I don't think this is safer; if folks are using their own regexes in `:jar-exclusions` they are probably assuming the directory would be part of it the same way it is with things like `.gitignore`; it would be better to build a regex based on the path separator.

**Comment by brunchboy** [comment]
Yes, that was what I ended up concluding myself too. Oops! I forgot to reference this issue in my pull request, so l’l link them this way: #2470

**Comment by technomancy** [comment]
Should be fixed in #2470.
